### PR TITLE
Implement resample method

### LINKED ISSUE #11: Implementing the TimeIndex class - the resample method
We start to implement the resample method in `TimeIndex` class. For reference, the relevant legacy codes are:

- Aggregate the daily or monthly dataset to target day or month means
https://github.com/AI4S2S/proto/blob/77734930a40b8aaefcf1e390efe0e3ac93b40858/RGCPD/functions_pp.py#L501
- Mean for the entire period
https://github.com/AI4S2S/proto/blob/77734930a40b8aaefcf1e390efe0e3ac93b40858/RGCPD/functions_pp.py#L911

Our previous discussion about the design of this module can be found in https://github.com/AI4S2S/proto/issues/3.

### LINKED ISSUE #24: New name for column 'lag' in function map_to_data in time.py
We need a new name for the column name 'lag' in the function map_to_data. https://github.com/AI4S2S/ai4s2s/blob/fa0ad6508b7c1f9e294adfde527c786e7a8f7b8c/s2s/time.py#L176
The name lag does not quite cut it, because it does not acknowledge that the first period after the next target period is also lag 1,2,3 etc.
Current suggestions:
- interval_i
- count
- intval
- i
- bin

### PR DESCRIPTION
**Pull request by geek-yang**
In this PR, the `resample` method in `time.py` is implemented. It resamples input data to target frequency.

Example about how to use it:
```py
import s2s.time
import pandas as pd
import numpy as np

cal = s2s.time.AdventCalendar()
time_index = pd.date_range('20211101', '20211111', freq='1d')
var = np.arange(len(time_index))
input_data = pd.Series(var, index=time_index)
bins = cal.resample(input_data, target_freq='5d')
bins
>>> 2021-11-01     2.0
>>> 2021-11-06     7.0
>>> 2021-11-11    10.0
>>> Freq: 5D, dtype: float64  
```

A tutorial notebook is also included to explain how to use the function. For the future developments, we could always include notebooks in the `notebooks` folder to explain the new functions.

This PR closes #11, closes #24.

### CODE REVIEW

**Comment by geek-yang** [comment]
Note that currently the `resample` function only supports `pd.DataFrame`. We will extend it to support `xarray.DataArray` in another PR. I have created an issue #17 for this.

**Comment by Peter9192** [comment]
> A tutorial notebook is also included to explain how to use the function. For the future developments, we could always include notebooks in the `notebooks` folder to explain the new functions.

Would this come in addition to the example code in the method docstring? I like the idea of using notebooks as a user-guide in the documentation, in that case we could use [nbsphynx](https://nbsphinx.readthedocs.io/en/0.8.9/). But the downside is that, in my experience, maintaining those is a pain.

**Review by Peter9192 — COMMENTED** [review]
Nice work, especially for a first (substantial) contribution! I have a couple of comments on the code, perhaps we can have a look together. Also, the PR is a bit messed up, since it incorporated some of the abandoned #18. I fear there might be some conflicts and unclear history. Maybe the cleanest solution is to open a new PR, if it's not too much trouble.

File: s2s/time.py [hunk 1]
```diff
@@ -170,6 +173,199 @@ def map_years(
...
+    def resample(self, input_data: Union[pd.Series, pd.DataFrame, xr.Dataset, xr.DataArray]
+        ) -> Union[pd.DataFrame, xr.Dataset]:
+        """Resample input data to the calendar frequency.
+
+        Pass a pandas Series/DataFrame with a datetime axis, or an
+        xarray DataArray/Dataset with a datetime coordinate called 'time'.
+        It will return the same object with the datetimes resampled onto
+        this DateTimeIndex by calculating the mean of each bins.
+
+        Note that this function is intended for upscaling operations, which means
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply s2s/time.py:265]: Good to know! I'll try to keep this in mind.

File: s2s/time.py [hunk 2]
```diff
@@ -170,6 +173,199 @@ def map_years(
...
+            >>> import s2s.time
+            >>> import pandas as pd
+            >>> import numpy as np
+            >>> cal = s2s.time.AdventCalendar(freq='180d')
+            >>> time_index = pd.date_range('20191201', '20211231', freq='1d')
+            >>> var = np.arange(len(time_index))
+            >>> input_data = pd.Series(var, index=time_index)
+            >>> bins = cal.resample(input_data)
+            >>> bins
+                  anchor_year  lag                  interval   mean
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply s2s/time.py:296]: Yes, this has been renamed to i_interval

File: s2s/time.py [hunk 3]
```diff
@@ -170,6 +173,199 @@ def map_years(
...
+                2        2021    0  (2021-06-03, 2021-11-30]  640.5
+                3        2021    1  (2020-12-05, 2021-06-03]  460.5
+        """
+        if isinstance(input_data, (pd.Series, pd.DataFrame)):
+            if not isinstance(input_data.index, pd.DatetimeIndex):
+                raise ValueError('The input data does not have a datetime index.')
+
+            # raise a warning for upscaling
+            # check if the time index of input data is reverse
+            if "-" in input_data.index.freqstr:
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply s2s/time.py:308]: It does not, so it is indeed good to check if it exists. I also rewrote this check to be more clear and concise.

File: s2s/time.py [hunk 4]
```diff
@@ -170,6 +173,199 @@ def map_years(
...
+                # target frequency must be larger than the original frequency
+                if pd.Timedelta(self.freq) < -input_data.index.freq:
+                    warnings.warn("Target frequency is smaller than the original frequency."
+                        + "It is upscaling and please check the returned values.")
+            else:
+                if pd.Timedelta(self.freq) < input_data.index.freq:
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply s2s/time.py:314]: I refactored the entire resampling function, it should be much more clear (just waiting for the map_to_data PR to finish.)

File: tests/test_time.py [hunk 5]
```diff
@@ -58,6 +58,129 @@ def test_map_years(self):
...
+        timeseries = pd.Series(test_data, index=time_index)
+        
+        # test pandas Series without name
+        resampled_data = cal.resample(timeseries)
+        expected = np.array([test_data[4:7].mean(), test_data[1:4].mean()])
+
+        assert np.array_equal(resampled_data['mean_data'].values, expected)
+
+        # test pandas Series with name
+        timeseries = timeseries.rename('data1')
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply tests/test_time.py:126]: I have done this, and included pytest's parameterized fixtures to make the testing more clean and compact

**Pull request closed without merging**